File: leetcode/540.single-element-in-a-sorted-array.py
# @leet start
class Solution:
    def singleNonDuplicate(self, nums: list[int]) -> int:
        # Edge case: only one element
        n = len(nums)

        if n == 1:
            return nums[0]

        # Now we do a binary search on all odd indices smallest odd
        # index i that satisfies nums[i] != nums[i - 1]. Using this i,
        # nums[i - 1] is the sole single element.
        odd_idxs = list(range(1, n, 2))

        # NOTE: in the binary search, left, right, and mid refer to
        # indices of the odd_idxs list
        num_odd_idxs = len(odd_idxs)

        left = 0
        right = num_odd_idxs - 1

        while left <= right:
            mid = (left + right) // 2

            odd_idx = odd_idxs[mid]

            if nums[odd_idx] != nums[odd_idx - 1]:
                # We've already passed the sole single element
                right = mid - 1
            else:
                # We haven't passed the sole single element
                left = mid + 1

        # Return the answer
        if left == num_odd_idxs:
            # There was no odd index satisfying the condition, so the
            # last element is the sole single element
            return nums[-1]

        smallest_odd_idx_after_single_element = odd_idxs[left]

        return nums[smallest_odd_idx_after_single_element - 1]


# @leet end

File: leetcode/540.single-element-in-a-sorted-array.js
// @leet start
/**
 * @param {number[]} nums
 * @return {number}
 */
var singleNonDuplicate = function (nums) {
  let left = 0;
  let right = nums.length - 1;

  while (left <= right) {
    const mid = left + Math.floor((right - left) / 2);

    if (
      (mid % 2 === 0 &&
        mid !== nums.length - 1 &&
        nums[mid] === nums[mid + 1]) ||
      (mid % 2 === 1 && nums[mid] === nums[mid - 1])
    ) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return nums[left];
};
// @leet end
